utils/lib/DataSource.ts

```
@@ -0,0 +1,149 @@
+      install: this._parseInstall(),
+      categoryTerms: categoryTerms && categoryTerms.map((t) => t.trim()),
+      keywords: keywords && keywords.map((k) => k.trim()),
+      description: description && description.trim(),
+    };
+
+    return {
+      dryRun,
+      id,
+      dataSourceMetadata,
+    };
+  }
+
+  public async submitMutation(dryRun = true) {
+    const { data, errors } = await fetchNRGraphqlResults<
+      DataSourceMutationVariable,
+      DataSourceMutationResponse
+    >({
+      queryString: DATA_SOURCE_MUTATION,
+      variables: this._getComponentMutationVariables(dryRun),
+    });
+
+    // filePath may need to be changed for this rework
+    return { data, errors, name: this.localPath };
+  }
+
+  /**
+   * Helper method to get the image URL for the icon.
+   */
+  private _getIconUrl() {
+    const { icon } = this.config;
+    const dirName = path.dirname(this.configPath);
+    const relDirName = removeBasePath(dirName, this.basePath);
+
+    return `${GITHUB_RAW_BASE_URL}/${relDirName}/${icon}`;
+  }
+
+  /**
+   * Helper method to parse the primary and fallback install directives.
+   */
+  private _parseInstall() {
+    const { install } = this.config;
+
+    return {
+      primary: this._parseInstallDirective(install.primary),
+      fallback:
+        install.fallback && this._parseInstallDirective(install.fallback),
+    };
+  }
+
+  /**
+   * Helper method that returns the directive, based on it's type.
+   */
+  private _parseInstallDirective(
+    directive: DataSourceConfigInstallDirective
+  ): DataSourceInstallDirectiveInput {
+    if ('link' in directive) {
+      const { url } = directive.link;
+
+      return {
+        link: {
+          url: url?.trim() ?? '',
+        },
+      };
+    }
+
+    if ('nerdlet' in directive) {
+      const { nerdletId, nerdletState, requiresAccount } = directive.nerdlet;
+
+      return {
+        nerdlet: {
+          nerdletId: nerdletId.trim() ?? '',
```

This should throw a `cannot reference trim() of undefined` error if the `nerdletId` is `undefined`. We can use optional chaining to deal with this.
```suggestion
          nerdletId: nerdletId?.trim() ?? '',
```